File: API/models/expense.py
from sqlalchemy import (
    CheckConstraint,
    Column,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import relationship

from models.base import Base


class Expense(Base):
    """
    Database model for the expense
    """

    __tablename__ = "expenses"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=True)
    group_id = Column(Integer, ForeignKey("groups.id", ondelete="CASCADE"), nullable=True)
    title = Column(String(255))
    amount = Column(Float, nullable=False)
    description = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    category_id = Column(Integer, ForeignKey("categories.id", ondelete="CASCADE"), nullable=False)

    user = relationship("User", back_populates="expenses", passive_deletes=True)
    group = relationship("Group", back_populates="expenses", passive_deletes=True)
    category = relationship("Category", back_populates="expenses", passive_deletes=True)

    __table_args__ = (
        CheckConstraint(
            "((user_id IS NOT NULL AND group_id IS NULL) OR "
            "(user_id IS NOT NULL AND group_id IS NOT NULL))",
            name="chk_expenses_one_fk"
        ),
    )

    @property
    def split_amount(self) -> float:
        """
        Calculates amount per person if it is a group expense.
        """
        print(f"DEBUG: Group {self.group_id} has members: {[u.id for u in self.group.users]}")
        if self.group_id and self.group and self.group.users:
            member_count = len(self.group.users)
            if member_count > 0:
                return round(self.amount / member_count, 2)
        return self.amount

File: Mobile/androidApp/src/main/java/com/example/budgeting/android/data/model/Expense.kt
package com.example.budgeting.android.data.model

import com.squareup.moshi.Json

data class Expense(
    @Json(name = "id")
    val id: Int? = null,
    
    @Json(name = "user_id")
    var user_id: Int? = null,
    
    @Json(name = "group_id")
    val group_id: Int? = null,
    
    @Json(name = "title")
    val title: String,
    
    @Json(name = "category_id")
    val category_id: Int? = null,
    
    @Json(name = "category")
    val category: Category? = null,
    
    @Json(name = "amount")
    val amount: Double,
    
    @Json(name = "description")
    val description: String? = null,
    
    @Json(name = "created_at")
    val created_at: String? = null
) {
    val categoryTitle: String
        get() = category?.title ?: "Uncategorized"
}

data class ExpenseIdResponse(
    @Json(name = "id")
    val id: Int
)

data class ExpenseCreateRequest(
    @Json(name = "title")
    val title: String,
    
    @Json(name = "amount")
    val amount: Double,
    
    @Json(name = "category_id")
    val category_id: Int,
    
    @Json(name = "group_id")
    val group_id: Int? = null,
    
    @Json(name = "description")
    val description: String? = null
)